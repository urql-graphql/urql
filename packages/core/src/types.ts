import { TypedDocumentNode } from '@graphql-typed-document-node/core';
import { GraphQLError, DocumentNode } from 'graphql';
import { Source } from 'wonka';
import { Client } from './client';
import { CombinedError } from './utils/error';

export type ExecutionResult =
  | {
      errors?:
        | Array<Partial<GraphQLError> | string | Error>
        | readonly GraphQLError[];
      data?: null | Record<string, any>;
      extensions?: Record<string, any>;
      hasNext?: boolean;
    }
  | {
      errors?:
        | Array<Partial<GraphQLError> | string | Error>
        | readonly GraphQLError[];
      data: any;
      path: (string | number)[];
      hasNext?: boolean;
    };

export type PromisifiedSource<T = any> = Source<T> & {
  toPromise: () => Promise<T>;
};

/** The type of GraphQL operation being executed. */
export type OperationType = 'subscription' | 'query' | 'mutation' | 'teardown';

/** The strategy that is used to request results from network and/or the cache. */
export type RequestPolicy =
  | 'cache-first'
  | 'cache-only'
  | 'network-only'
  | 'cache-and-network';

/** How the operation has */
export type CacheOutcome = 'miss' | 'partial' | 'hit';

/** A default type for variables */
export type AnyVariables = { [prop: string]: any } | void | undefined;

/** A Graphql query, mutation, or subscription. */
export interface GraphQLRequest<
  Data = any,
  Variables extends AnyVariables = AnyVariables
> {
  /** Unique identifier of the request. */
  key: number;
  query: DocumentNode | TypedDocumentNode<Data, Variables>;
  variables: Variables;
}

/** Metadata that is only available in development for devtools. */
export interface OperationDebugMeta {
  source?: string;
  cacheOutcome?: CacheOutcome;
  networkLatency?: number;
  startTime?: number;
}

/** A unique marker that marks the operation's identity when multiple mutation operations with identical keys are issued. */
export type OperationInstance = number & { readonly _opaque: unique symbol };

/** Additional metadata passed to [exchange]{@link Exchange} functions. */
export interface OperationContext {
  [key: string]: any;
  readonly _instance?: OperationInstance | undefined;
  additionalTypenames?: string[];
  fetch?: typeof fetch;
  fetchOptions?: RequestInit | (() => RequestInit);
  requestPolicy: RequestPolicy;
  url: string;
  meta?: OperationDebugMeta;
  suspense?: boolean;
  /** Instructs fetch exchanges to use a GET request.
   * When true or 'within-url-limit' is passed, GET won't be used when the resulting URL exceeds a length of 2048. */
  preferGetMethod?: boolean | 'force' | 'within-url-limit';
}

/** A [query]{@link Query} or [mutation]{@link Mutation} with additional metadata for use during transmission. */
export interface Operation<
  Data = any,
  Variables extends AnyVariables = AnyVariables
> extends GraphQLRequest<Data, Variables> {
  readonly kind: OperationType;
  context: OperationContext;
}

/** Resulting data from an [operation]{@link Operation}. */
export interface OperationResult<
  Data = any,
  Variables extends AnyVariables = AnyVariables
> {
  /** The [operation]{@link Operation} which has been executed. */
  operation: Operation<Data, Variables>;
  /** The data returned from the Graphql server. */
  data?: Data;
  /** Any errors resulting from the operation. */
  error?: CombinedError;
  /** Optional extensions return by the Graphql server. */
  extensions?: Record<string, any>;
  /** Optional stale flag added by exchanges that return stale results. */
  stale?: boolean;
  /** Optional hasNext flag indicating deferred/streamed results are following. */
  hasNext?: boolean;
}

/** Input parameters for to an Exchange factory function. */
export interface ExchangeInput {
  client: Client;
  forward: ExchangeIO;
  dispatchDebug: <T extends keyof DebugEventTypes | string>(
    t: DebugEventArg<T>
  ) => void;
}

/** Function responsible for listening for streamed [operations]{@link Operation}. */
export type Exchange = (input: ExchangeInput) => ExchangeIO;

/** Function responsible for receiving an observable [operation]{@link Operation} and returning a [result]{@link OperationResult}. */
export type ExchangeIO = (ops$: Source<Operation>) => Source<OperationResult>;

/** Debug event types (interfaced for declaration merging). */
export interface DebugEventTypes {
  // Cache exchange
  cacheHit: { value: any };
  cacheInvalidation: {
    typenames: string[];
    response: OperationResult;
  };
  // Fetch exchange
  fetchRequest: {
    url: string;
    fetchOptions: RequestInit;
  };
  fetchSuccess: {
    url: string;
    fetchOptions: RequestInit;
    value: object;
  };
  fetchError: {
    url: string;
    fetchOptions: RequestInit;
    value: Error;
  };
  // Retry exchange
  retryRetrying: {
    retryCount: number;
  };
}

export type DebugEventArg<T extends keyof DebugEventTypes | string> = {
  type: T;
  message: string;
  operation: Operation;
} & (T extends keyof DebugEventTypes
  ? { data: DebugEventTypes[T] }
  : { data?: any });

export type DebugEvent<
  T extends keyof DebugEventTypes | string = string
> = DebugEventArg<T> & {
  timestamp: number;
  source: string;
};
