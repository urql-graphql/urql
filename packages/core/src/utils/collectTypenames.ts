interface EntityLike {
  [key: string]: EntityLike | EntityLike[] | any;
  __typename: string | null | void;
}

const collectTypes = (obj: EntityLike | EntityLike[], types: Set<string>) => {
  if (Array.isArray(obj)) {
    for (const item of obj) collectTypes(item, types);
  } else if (typeof obj === 'object' && obj !== null) {
    for (const key in obj) {
      if (key === '__typename' && typeof obj[key] === 'string') {
        types.add(obj[key] as string);
      } else {
        collectTypes(obj[key], types);
      }
    }
  }

  return types;
};

/** Finds and returns a list of `__typename` fields found in response data.
 *
 * @privateRemarks
 * This is used by `@urql/core`’s document `cacheExchange` to find typenames
 * in a given GraphQL response’s data.
 */
export const collectTypenames = (response: object): string[] => [
  ...collectTypes(response as EntityLike, new Set()),
];
