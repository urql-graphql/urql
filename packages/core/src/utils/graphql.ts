import type * as GraphQLWeb from '@0no-co/graphql.web';
import type * as GraphQL from 'graphql';

type OrNever<T> = 0 extends 1 & T ? never : T;

export type GraphQLError =
  | GraphQLWeb.GraphQLError
  | OrNever<GraphQL.GraphQLError>;

export type DocumentNode =
  | GraphQLWeb.DocumentNode
  | OrNever<GraphQL.DocumentNode>;

export type DefinitionNode =
  | GraphQLWeb.DefinitionNode
  | OrNever<GraphQL.DefinitionNode>;
