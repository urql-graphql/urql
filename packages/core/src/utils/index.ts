export * from './error';
export * from './request';
export * from './result';
export * from './variables';
export * from './collectTypenames';
export * from './formatDocument';
export * from './maskTypename';
export * from './streamUtils';
export * from './operation';

export const noop = () => {
  /* noop */
};
