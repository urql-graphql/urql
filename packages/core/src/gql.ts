/* eslint-disable prefer-rest-params */
import { DocumentNode, DefinitionNode, Kind } from 'graphql';
import { TypedDocumentNode } from './types';
import { keyDocument, stringifyDocument } from './utils';

const applyDefinitions = (
  fragmentNames: Map<string, string>,
  target: DefinitionNode[],
  source: Array<DefinitionNode> | ReadonlyArray<DefinitionNode>
) => {
  for (const definition of source) {
    if (definition.kind === Kind.FRAGMENT_DEFINITION) {
      const name = definition.name.value;
      const value = stringifyDocument(definition);
      // Fragments will be deduplicated according to this Map
      if (!fragmentNames.has(name)) {
        fragmentNames.set(name, value);
        target.push(definition);
      } else if (
        process.env.NODE_ENV !== 'production' &&
        fragmentNames.get(name) !== value
      ) {
        // Fragments with the same names is expected to have the same contents
        console.warn(
          '[WARNING: Duplicate Fragment] A fragment with name `' +
            name +
            '` already exists in this document.\n' +
            'While fragment names may not be unique across your source, each name must be unique per document.'
        );
      }
    } else {
      target.push(definition);
    }
  }
};

function gql<Data = any, Variables = object>(
  strings: TemplateStringsArray,
  ...interpolations: Array<TypedDocumentNode | DocumentNode | string>
): TypedDocumentNode<Data, Variables>;

function gql<Data = any, Variables = object>(
  string: string
): TypedDocumentNode<Data, Variables>;

function gql(/* arguments */) {
  const fragmentNames = new Map<string, string>();
  const definitions: DefinitionNode[] = [];
  const interpolations: DefinitionNode[] = [];

  // Apply the entire tagged template body's definitions
  let body: string = Array.isArray(arguments[0])
    ? arguments[0][0]
    : arguments[0] || '';
  for (let i = 1; i < arguments.length; i++) {
    const value = arguments[i];
    if (value && value.definitions) {
      interpolations.push(...value.definitions);
    } else {
      body += value;
    }

    body += arguments[0][i];
  }

  // Apply the tag's body definitions
  applyDefinitions(fragmentNames, definitions, keyDocument(body).definitions);
  // Copy over each interpolated document's definitions
  applyDefinitions(fragmentNames, definitions, interpolations);

  return keyDocument({
    kind: Kind.DOCUMENT,
    definitions,
  });
}

export { gql };
