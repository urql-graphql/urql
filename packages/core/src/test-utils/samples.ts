import { gql } from '../gql';

import {
  ExecutionResult,
  GraphQLRequest,
  Operation,
  OperationContext,
  OperationResult,
} from '../types';
import { makeOperation } from '../utils';

export const context: OperationContext = {
  fetchOptions: {
    method: 'POST',
  },
  requestPolicy: 'cache-first',
  url: 'http://localhost:3000/graphql',
};

export const queryGql: GraphQLRequest = {
  key: 2,
  query: gql`
    query getUser($name: String) {
      user(name: $name) {
        id
        firstName
        lastName
      }
    }
  `,
  variables: {
    name: 'Clara',
  },
};

export const mutationGql: GraphQLRequest = {
  key: 3,
  query: gql`
    mutation AddUser($name: String) {
      addUser(name: $name) {
        name
      }
    }
  `,
  variables: {
    name: 'Clara',
  },
};

export const subscriptionGql: GraphQLRequest = {
  key: 4,
  query: gql`
    subscription subscribeToUser($user: String) {
      user(user: $user) {
        name
      }
    }
  `,
  variables: {
    user: 'colin',
  },
};

export const queryOperation: Operation = makeOperation(
  'query',
  {
    query: queryGql.query,
    variables: queryGql.variables,
    key: queryGql.key,
  },
  context
);

export const teardownOperation: Operation = makeOperation(
  'teardown',
  {
    query: queryOperation.query,
    variables: queryOperation.variables,
    key: queryOperation.key,
  },
  context
);

export const mutationOperation: Operation = makeOperation(
  'mutation',
  {
    query: mutationGql.query,
    variables: mutationGql.variables,
    key: mutationGql.key,
  },
  context
);

export const subscriptionOperation: Operation = makeOperation(
  'subscription',
  {
    query: subscriptionGql.query,
    variables: subscriptionGql.variables,
    key: subscriptionGql.key,
  },
  context
);

export const undefinedQueryResponse: OperationResult = {
  operation: queryOperation,
};

export const queryResponse: OperationResult = {
  operation: queryOperation,
  data: {
    user: {
      name: 'Clive',
    },
  },
};

export const mutationResponse: OperationResult = {
  operation: mutationOperation,
  data: {},
};

export const subscriptionResult: ExecutionResult = {
  data: {},
};
