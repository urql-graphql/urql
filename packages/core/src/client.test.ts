import { print } from 'graphql';
import { vi, expect, it, beforeEach, describe, afterEach } from 'vitest';

/** NOTE: Testing in this file is designed to test both the client and its interaction with default Exchanges */

import {
  Source,
  delay,
  map,
  never,
  pipe,
  merge,
  subscribe,
  publish,
  filter,
  share,
  toArray,
  toPromise,
  onPush,
  tap,
  take,
} from 'wonka';

import { gql } from './gql';
import { Exchange, Operation, OperationResult } from './types';
import { makeOperation } from './utils';
import { Client, createClient } from './client';
import { queryOperation, subscriptionOperation } from './test-utils';

const url = 'https://hostname.com';

describe('createClient / Client', () => {
  it('creates an instance of Client', () => {
    expect(createClient({ url }) instanceof Client).toBeTruthy();
    expect(new Client({ url }) instanceof Client).toBeTruthy();
  });

  it('passes snapshot', () => {
    const client = createClient({ url });
    expect(client).toMatchSnapshot();
  });
});

const query = {
  key: 1,
  query: gql`
    {
      todos {
        id
      }
    }
  `,
  variables: { example: 1234 },
};

const mutation = {
  key: 1,
  query: gql`
    mutation {
      todos {
        id
      }
    }
  `,
  variables: { example: 1234 },
};

const subscription = {
  key: 1,
  query: gql`
    subscription {
      todos {
        id
      }
    }
  `,
  variables: { example: 1234 },
};

let receivedOps: Operation[] = [];
let client = createClient({ url: '1234' });
const receiveMock = vi.fn((s: Source<Operation>) =>
  pipe(
    s,
    tap(op => (receivedOps = [...receivedOps, op])),
    map(op => ({ operation: op }))
  )
);
const exchangeMock = vi.fn(() => receiveMock);

beforeEach(() => {
  receivedOps = [];
  exchangeMock.mockClear();
  receiveMock.mockClear();
  client = createClient({
    url,
    exchanges: [exchangeMock] as any[],
    requestPolicy: 'cache-and-network',
  });
});

describe('exchange args', () => {
  it('receives forward function', () => {
    // @ts-ignore
    expect(typeof exchangeMock.mock.calls[0][0].forward).toBe('function');
  });

  it('receives client', () => {
    // @ts-ignore
    expect(exchangeMock.mock.calls[0][0]).toHaveProperty('client', client);
  });
});

describe('promisified methods', () => {
  it('query', () => {
    const queryResult = client
      .query(
        gql`
          {
            todos {
              id
            }
          }
        `,
        { example: 1234 },
        { requestPolicy: 'cache-only' }
      )
      .toPromise();

    const received = receivedOps[0];
    expect(print(received.query)).toEqual(print(query.query));
    expect(received.key).toBeDefined();
    expect(received.variables).toEqual({ example: 1234 });
    expect(received.kind).toEqual('query');
    expect(received.context).toEqual({
      url: 'https://hostname.com',
      requestPolicy: 'cache-only',
      fetchOptions: undefined,
      fetch: undefined,
      suspense: false,
      preferGetMethod: false,
    });
    expect(queryResult).toHaveProperty('then');
  });

  it('mutation', () => {
    const mut = gql`
      mutation {
        todos {
          id
        }
      }
    `;
    const mutationResult = client.mutation(mut, { example: 1234 }).toPromise();

    const received = receivedOps[0];
    expect(print(received.query)).toEqual(print(mut));
    expect(received.key).toBeDefined();
    expect(received.variables).toEqual({ example: 1234 });
    expect(received.kind).toEqual('mutation');
    expect(received.context).toMatchObject({
      url: 'https://hostname.com',
      requestPolicy: 'cache-and-network',
      fetchOptions: undefined,
      fetch: undefined,
      suspense: false,
      preferGetMethod: false,
    });
    expect(mutationResult).toHaveProperty('then');
  });
});

describe('synchronous methods', () => {
  it('readQuery', () => {
    const result = client.readQuery(
      gql`
        {
          todos {
            id
          }
        }
      `,
      { example: 1234 }
    );

    expect(receivedOps.length).toBe(2);
    expect(receivedOps[0].kind).toBe('query');
    expect(receivedOps[1].kind).toBe('teardown');
    expect(result).toEqual({
      operation: {
        ...query,
        context: expect.anything(),
        key: expect.any(Number),
        kind: 'query',
      },
    });
  });
});

describe('executeQuery', () => {
  it('passes query string exchange', () => {
    pipe(
      client.executeQuery(query),
      subscribe(x => x)
    );

    const receivedQuery = receivedOps[0].query;
    expect(print(receivedQuery)).toBe(print(query.query));
  });

  it('should throw when passing in a mutation', () => {
    try {
      client.executeQuery(mutation);
      expect(true).toBeFalsy();
    } catch (e: any) {
      expect(e.message).toMatchInlineSnapshot(
        `"Expected operation of type \\"query\\" but found \\"mutation\\""`
      );
    }
  });

  it('passes variables type to exchange', () => {
    pipe(
      client.executeQuery(query),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('variables', query.variables);
  });

  it('passes requestPolicy to exchange', () => {
    pipe(
      client.executeQuery(query),
      subscribe(x => x)
    );

    expect(receivedOps[0].context).toHaveProperty(
      'requestPolicy',
      'cache-and-network'
    );
  });

  it('allows overriding the requestPolicy', () => {
    pipe(
      client.executeQuery(query, { requestPolicy: 'cache-first' }),
      subscribe(x => x)
    );

    expect(receivedOps[0].context).toHaveProperty(
      'requestPolicy',
      'cache-first'
    );
  });

  it('passes kind type to exchange', () => {
    pipe(
      client.executeQuery(query),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('kind', 'query');
  });

  it('passes url (from context) to exchange', () => {
    pipe(
      client.executeQuery(query),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('context.url', url);
  });
});

describe('executeMutation', () => {
  it('passes query string exchange', async () => {
    pipe(
      client.executeMutation(mutation),
      subscribe(x => x)
    );

    const receivedQuery = receivedOps[0].query;
    expect(print(receivedQuery)).toBe(print(mutation.query));
  });

  it('passes variables type to exchange', () => {
    pipe(
      client.executeMutation(mutation),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('variables', query.variables);
  });

  it('passes kind type to exchange', () => {
    pipe(
      client.executeMutation(mutation),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('kind', 'mutation');
  });

  it('passes url (from context) to exchange', () => {
    pipe(
      client.executeMutation(mutation),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('context.url', url);
  });
});

describe('executeSubscription', () => {
  it('passes query string exchange', async () => {
    pipe(
      client.executeSubscription(subscription),
      subscribe(x => x)
    );

    const receivedQuery = receivedOps[0].query;
    expect(print(receivedQuery)).toBe(print(subscription.query));
  });

  it('passes variables type to exchange', () => {
    pipe(
      client.executeSubscription(subscription),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('variables', subscription.variables);
  });

  it('passes kind type to exchange', () => {
    pipe(
      client.executeSubscription(subscription),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('kind', 'subscription');
  });
});

describe('queuing behavior', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('queues reexecuteOperation, which dispatchOperation consumes', () => {
    const output: Array<Operation | OperationResult> = [];

    const exchange: Exchange = ({ client }) => ops$ => {
      return pipe(
        ops$,
        filter(op => op.kind !== 'teardown'),
        tap(op => {
          output.push(op);
          if (
            op.key === queryOperation.key &&
            op.context.requestPolicy === 'cache-first'
          ) {
            client.reexecuteOperation({
              ...op,
              context: {
                ...op.context,
                requestPolicy: 'network-only',
              },
            });
          }
        }),
        map(op => ({
          data: op.key,
          operation: op,
        }))
      );
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const shared = pipe(
      client.executeRequestOperation(queryOperation),
      onPush(result => output.push(result)),
      share
    );

    const results = pipe(shared, toArray);
    pipe(shared, publish);

    expect(output.length).toBe(8);
    expect(results.length).toBe(2);

    expect(output[0]).toHaveProperty('key', queryOperation.key);
    expect(output[0]).toHaveProperty('context.requestPolicy', 'cache-first');

    expect(output[1]).toHaveProperty('operation.key', queryOperation.key);
    expect(output[1]).toHaveProperty(
      'operation.context.requestPolicy',
      'cache-first'
    );

    expect(output[2]).toHaveProperty('key', queryOperation.key);
    expect(output[2]).toHaveProperty('context.requestPolicy', 'network-only');

    expect(output[3]).toHaveProperty('operation.key', queryOperation.key);
    expect(output[3]).toHaveProperty(
      'operation.context.requestPolicy',
      'network-only'
    );

    expect(output[1]).toBe(results[0]);
    expect(output[3]).toBe(results[1]);
  });

  it('reemits previous results as stale if the operation is reexecuted as network-only', async () => {
    const output: OperationResult[] = [];

    const exchange: Exchange = () => {
      let countRes = 0;
      return ops$ => {
        return pipe(
          ops$,
          filter(op => op.kind !== 'teardown'),
          map(op => ({
            data: ++countRes,
            operation: op,
          })),
          delay(1)
        );
      };
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const { unsubscribe } = pipe(
      client.executeRequestOperation(queryOperation),
      subscribe(result => {
        output.push(result);
      })
    );

    vi.advanceTimersByTime(1);

    expect(output.length).toBe(1);
    expect(output[0]).toHaveProperty('data', 1);
    expect(output[0]).toHaveProperty('operation.key', queryOperation.key);
    expect(output[0]).toHaveProperty(
      'operation.context.requestPolicy',
      'cache-first'
    );

    client.reexecuteOperation(
      makeOperation(queryOperation.kind, queryOperation, {
        ...queryOperation.context,
        requestPolicy: 'network-only',
      })
    );

    await Promise.resolve();

    expect(output.length).toBe(2);
    expect(output[1]).toHaveProperty('data', 1);
    expect(output[1]).toHaveProperty('stale', true);
    expect(output[1]).toHaveProperty('operation.key', queryOperation.key);
    expect(output[1]).toHaveProperty(
      'operation.context.requestPolicy',
      'cache-first'
    );

    vi.advanceTimersByTime(1);

    expect(output.length).toBe(3);
    expect(output[2]).toHaveProperty('data', 2);
    expect(output[2]).not.toHaveProperty('stale');
    expect(output[2]).toHaveProperty('operation.key', queryOperation.key);
    expect(output[2]).toHaveProperty(
      'operation.context.requestPolicy',
      'network-only'
    );

    unsubscribe();
  });

  it('does not reemit previous results as stale if it was marked as stale already', async () => {
    const output: OperationResult[] = [];
    const exchange: Exchange = () => ops$ => {
      return pipe(
        ops$,
        filter(op => op.kind !== 'teardown'),
        map(op => ({
          data: 1,
          operation: op,
          stale: true,
        })),
        delay(1)
      );
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const { unsubscribe } = pipe(
      client.executeRequestOperation(queryOperation),
      subscribe(result => {
        output.push(result);
      })
    );

    vi.advanceTimersByTime(1);

    expect(output.length).toBe(1);
    expect(output[0]).toHaveProperty('operation.key', queryOperation.key);
    expect(output[0]).toHaveProperty(
      'operation.context.requestPolicy',
      'cache-first'
    );

    client.reexecuteOperation(
      makeOperation(queryOperation.kind, queryOperation, {
        ...queryOperation.context,
        requestPolicy: 'network-only',
      })
    );

    await Promise.resolve();
    vi.advanceTimersByTime(1);

    expect(output.length).toBe(2);
    expect(output[1]).toHaveProperty('stale', true);
    expect(output[1]).toHaveProperty('operation.key', queryOperation.key);
    expect(output[1]).toHaveProperty(
      'operation.context.requestPolicy',
      'network-only'
    );

    unsubscribe();
  });
});

describe('shared sources behavior', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('replays results from prior operation result as needed (cache-first)', async () => {
    const exchange: Exchange = () => ops$ => {
      let i = 0;
      return pipe(
        ops$,
        map(op => ({
          data: ++i,
          operation: op,
        })),
        delay(1)
      );
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const resultOne = vi.fn();
    const resultTwo = vi.fn();

    pipe(client.executeRequestOperation(queryOperation), subscribe(resultOne));

    expect(resultOne).toHaveBeenCalledTimes(0);

    vi.advanceTimersByTime(1);

    expect(resultOne).toHaveBeenCalledTimes(1);
    expect(resultOne).toHaveBeenCalledWith({
      data: 1,
      operation: queryOperation,
    });

    pipe(client.executeRequestOperation(queryOperation), subscribe(resultTwo));

    expect(resultTwo).toHaveBeenCalledWith({
      data: 1,
      operation: queryOperation,
    });

    vi.advanceTimersByTime(1);

    // With cache-first we don't expect a new operation to be issued
    expect(resultTwo).toHaveBeenCalledTimes(1);
  });

  it('dispatches the correct request policy on subsequent sources', async () => {
    const exchange: Exchange = () => ops$ => {
      let i = 0;
      return pipe(
        ops$,
        map(op => ({
          data: ++i,
          operation: op,
        })),
        delay(1)
      );
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const resultOne = vi.fn();
    const resultTwo = vi.fn();
    const operationOne = makeOperation('query', queryOperation, {
      ...queryOperation.context,
      requestPolicy: 'cache-first',
    });
    const operationTwo = makeOperation('query', queryOperation, {
      ...queryOperation.context,
      requestPolicy: 'network-only',
    });

    pipe(client.executeRequestOperation(operationOne), subscribe(resultOne));

    expect(resultOne).toHaveBeenCalledTimes(0);

    vi.advanceTimersByTime(1);

    expect(resultOne).toHaveBeenCalledTimes(1);
    expect(resultOne).toHaveBeenCalledWith({
      data: 1,
      operation: operationOne,
    });

    pipe(client.executeRequestOperation(operationTwo), subscribe(resultTwo));

    expect(resultTwo).toHaveBeenCalledWith({
      data: 1,
      operation: operationOne,
      stale: true,
    });

    vi.advanceTimersByTime(1);

    expect(resultTwo).toHaveBeenCalledWith({
      data: 2,
      operation: operationTwo,
    });
  });

  it('replays results from prior operation result as needed (network-only)', async () => {
    const exchange: Exchange = () => ops$ => {
      let i = 0;
      return pipe(
        ops$,
        map(op => ({
          data: ++i,
          operation: op,
        })),
        delay(1)
      );
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const operation = makeOperation('query', queryOperation, {
      ...queryOperation.context,
      requestPolicy: 'network-only',
    });

    const resultOne = vi.fn();
    const resultTwo = vi.fn();

    pipe(client.executeRequestOperation(operation), subscribe(resultOne));

    expect(resultOne).toHaveBeenCalledTimes(0);

    vi.advanceTimersByTime(1);

    expect(resultOne).toHaveBeenCalledTimes(1);
    expect(resultOne).toHaveBeenCalledWith({
      data: 1,
      operation,
    });

    pipe(client.executeRequestOperation(operation), subscribe(resultTwo));

    expect(resultTwo).toHaveBeenCalledWith({
      data: 1,
      operation,
      stale: true,
    });

    vi.advanceTimersByTime(1);

    // With network-only we expect a new operation to be issued, hence a new result
    expect(resultTwo).toHaveBeenCalledTimes(2);

    expect(resultTwo).toHaveBeenCalledWith({
      data: 2,
      operation,
    });
  });

  it('does not replay values from a past subscription', async () => {
    const exchange: Exchange = () => ops$ => {
      let i = 0;
      return pipe(
        ops$,
        filter(op => op.kind !== 'teardown'),
        map(op => ({
          data: ++i,
          operation: op,
        })),
        delay(1)
      );
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    // We keep the source in-memory
    const source = client.executeRequestOperation(queryOperation);
    const resultOne = vi.fn();
    let subscription;

    subscription = pipe(source, subscribe(resultOne));

    expect(resultOne).toHaveBeenCalledTimes(0);
    vi.advanceTimersByTime(1);

    expect(resultOne).toHaveBeenCalledWith({
      data: 1,
      operation: queryOperation,
    });

    subscription.unsubscribe();
    const resultTwo = vi.fn();
    subscription = pipe(source, subscribe(resultTwo));

    expect(resultTwo).toHaveBeenCalledTimes(0);
    vi.advanceTimersByTime(1);

    expect(resultTwo).toHaveBeenCalledWith({
      data: 2,
      operation: queryOperation,
    });
  });

  it('replayed results are not emitted on the shared source', () => {
    const exchange: Exchange = () => ops$ => {
      let i = 0;
      return pipe(
        ops$,
        map(op => ({
          data: ++i,
          operation: op,
        })),
        take(1)
      );
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const operation = makeOperation('query', queryOperation, {
      ...queryOperation.context,
      requestPolicy: 'network-only',
    });

    const resultOne = vi.fn();
    const resultTwo = vi.fn();

    pipe(client.executeRequestOperation(operation), subscribe(resultOne));
    pipe(client.executeRequestOperation(operation), subscribe(resultTwo));

    expect(resultTwo).toHaveBeenCalledTimes(1);
    expect(resultTwo).toHaveBeenCalledWith({
      data: 1,
      operation,
      stale: true,
    });
  });

  it('does nothing when no operation result has been emitted yet', () => {
    const exchange: Exchange = () => ops$ => {
      return pipe(
        ops$,
        map(op => ({ data: 1, operation: op })),
        filter(() => false)
      );
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const resultOne = vi.fn();
    const resultTwo = vi.fn();

    pipe(client.executeRequestOperation(queryOperation), subscribe(resultOne));

    pipe(client.executeRequestOperation(queryOperation), subscribe(resultTwo));

    expect(resultOne).toHaveBeenCalledTimes(0);
    expect(resultTwo).toHaveBeenCalledTimes(0);
  });

  it('skips replaying results when a result is emitted immediately (network-only)', () => {
    const exchange: Exchange = () => ops$ => {
      let i = 0;
      return pipe(
        ops$,
        map(op => ({ data: ++i, operation: op }))
      );
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const operation = makeOperation('query', queryOperation, {
      ...queryOperation.context,
      requestPolicy: 'network-only',
    });

    const resultOne = vi.fn();
    const resultTwo = vi.fn();

    pipe(client.executeRequestOperation(operation), subscribe(resultOne));

    expect(resultOne).toHaveBeenCalledWith({
      data: 1,
      operation,
    });

    pipe(client.executeRequestOperation(operation), subscribe(resultTwo));

    expect(resultTwo).toHaveBeenCalledWith({
      data: 2,
      operation,
    });

    expect(resultOne).toHaveBeenCalledWith({
      data: 2,
      operation,
    });
  });

  it('replays stale results as needed', () => {
    const exchange: Exchange = () => ops$ => {
      return pipe(
        ops$,
        map(op => ({ stale: true, data: 1, operation: op })),
        take(1)
      );
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const resultOne = vi.fn();
    const resultTwo = vi.fn();

    pipe(client.executeRequestOperation(queryOperation), subscribe(resultOne));

    expect(resultOne).toHaveBeenCalledWith({
      data: 1,
      operation: queryOperation,
      stale: true,
    });

    pipe(client.executeRequestOperation(queryOperation), subscribe(resultTwo));

    expect(resultTwo).toHaveBeenCalledWith({
      data: 1,
      operation: queryOperation,
      stale: true,
    });
  });

  it('does nothing when operation is a subscription has been emitted yet', () => {
    const exchange: Exchange = () => ops$ => {
      return merge([
        pipe(
          ops$,
          map(op => ({ data: 1, operation: op })),
          take(1)
        ),
        never,
      ]);
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const resultOne = vi.fn();
    const resultTwo = vi.fn();

    pipe(
      client.executeRequestOperation(subscriptionOperation),
      subscribe(resultOne)
    );
    expect(resultOne).toHaveBeenCalledTimes(1);

    pipe(
      client.executeRequestOperation(subscriptionOperation),
      subscribe(resultTwo)
    );
    expect(resultTwo).toHaveBeenCalledTimes(0);
  });

  it('supports promisified sources', async () => {
    const exchange: Exchange = () => ops$ => {
      return pipe(
        ops$,
        map(op => ({ stale: true, data: 1, operation: op }))
      );
    };

    const client = createClient({
      url: 'test',
      exchanges: [exchange],
    });

    const resultOne = await pipe(
      client.executeRequestOperation(queryOperation),
      take(1),
      toPromise
    );

    expect(resultOne).toEqual({
      data: 1,
      operation: queryOperation,
      stale: true,
    });
  });
});
