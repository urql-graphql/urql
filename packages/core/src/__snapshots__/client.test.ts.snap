// Vitest Snapshot v1

exports[`createClient / Client > passes snapshot 1`] = `
Client2 {
  "createRequestOperation": [Function],
  "executeMutation": [Function],
  "executeQuery": [Function],
  "executeRequestOperation": [Function],
  "executeSubscription": [Function],
  "mutation": [Function],
  "operations$": [Function],
  "query": [Function],
  "readQuery": [Function],
  "reexecuteOperation": [Function],
  "subscribeToDebugTarget": [Function],
  "subscription": [Function],
  "suspense": false,
}
`;
